src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -0,0 +1,119 @@
+    uint8 internal constant ORDER_FILL_BIT = 1;
+
+    /// @dev Decodes a signed order from calldata bytes into memory.
+    /// Orders are tightly packed in order to reduce calldata and associated gas
+    /// costs. They contain the following fields:
+    /// ```
+    /// struct Order {
+    ///     uint8 sellTokenId;
+    ///     uint8 buyTokenId;
+    ///     uint256 sellAmount;
+    ///     uint256 buyAmount;
+    ///     uint32 validTo;
+    ///     uint256 tip;
+    ///     uint8 flags;
+    ///     uint256 executedAmount;
+    ///     uint32 nonce;
+    ///     Signature {
+    ///         uint8 v;
+    ///         bytes32 r;
+    ///         bytes32 s;
+    ///     } signature;
+    /// }
+    /// ```
+    /// @param tokens The list of tokens included in the settlement. Token IDs
+    /// represent indices in this array to actual addresses.
+    /// @param encodedOrder The order as encoded calldata bytes.
+    /// @param order The memory location to decode the order to.
+    function decodeSignedOrder(
+        IERC20[] calldata tokens,
+        bytes calldata encodedOrder,
+        bytes32 domainSeparator,
+        Order memory order
+    ) internal pure {
+        // NOTE: This is currently unnecessarily gas inefficient. Specifically,
+        // there is a potentially extraneous check to the encoded order length
+        // (this can be verified once for the total encoded orders length).
+        // Additionally, Solidity generates bounds checks for each `abi.decode`
+        // and slice operation. Unfortunately using `assmebly { calldataload }`
+        // is quite ugly here since there is no mechanism to get calldata
+        // offsets (like there is for memory offsets) without manual
+        // computation, which is brittle as changes to the calling function
+        // signature would require manual adjustments to the computation. Once
+        // gas benchmarking is set up, we can evaluate if it is worth the extra
+        // effort.
+
+        require(encodedOrder.length == ORDER_STRIDE, "malformed order data");
+
+        order.sellToken = tokens[uint256(uint8(encodedOrder[0]))];
+        order.buyToken = tokens[uint256(uint8(encodedOrder[1]))];
+        order.sellAmount = abi.decode(encodedOrder[2:], (uint256));
+        order.buyAmount = abi.decode(encodedOrder[34:], (uint256));
+        order.validTo = uint32(
+            abi.decode(encodedOrder[66:], (uint256)) >> (256 - 32)
+        );
+        order.tip = abi.decode(encodedOrder[70:], (uint256));
+        uint8 flags = uint8(encodedOrder[102]);
+        order.executedAmount = abi.decode(encodedOrder[103:], (uint256));
+        uint32 nonce = uint32(
+            abi.decode(encodedOrder[135:], (uint256)) >> (256 - 32)
+        );
+        uint8 v = uint8(encodedOrder[139]);
+        bytes32 r = abi.decode(encodedOrder[140:], (bytes32));
+        bytes32 s = abi.decode(encodedOrder[172:], (bytes32));
+
+        bytes32 digest = keccak256(
```

Comment:
You are separating this encoding nicely into a library etc. Just like a well trained, clean coder would do.

The issue is that we want to reuse this digest later for the key of the bitmap and for efficiency we don't want to calculate it twice.
One solution would be to return the digest. 

But I fear the most gas-efficient way would not factor this encoding into a pure library function, but rather have a convoluted logic that sets the bitmap[key] and does the decoding at once :)

Reply:
> One solution would be to return the digest.

I thought of this after I opened the PR, I will add it :smile:.

> But I fear the most gas-efficient way would not factor this encoding into a pure library function, but rather have a convoluted logic that sets the bitmap[key] and does the decoding at once :)

I think if we return the digest, it just gets put on the stack, so it won't increase the total amount of memory needed and should be gas efficient as well.

Reply:
If you are later on calling your function decodeSignedOrder  from another function that splits the bytes calldata encodedOrderData into their data packages, then  you have to:
1) Either collect all the digests in there and return them, such that another function can look at all the digests and mark the nonce
2) Or use each single digest and call within the "orderDataSplitting function" another function for mark the nonce

Solution 1 would be the cleanest regarding the code. Solution 2 would be more gas efficient. 

If we are deciding for solution 2, then we can do the marking of the nonce even in this function: `decodeSignedOrder` as we are going for an unclean solution anyways.
Do you prefer solution 1 ?

Reply:
Personally, I'm a fan of solution 2.

The way I had originally planned to make the settlement as gas efficient as possible was to process each order one at a time while decoding. This would allow us to not collect a large order array but instead collect only the final `(token: address, recipient: address, executedSellAmount: uint256, executedBuyAmount: uint256)` tuple that we need for transferring the sell amounts to the settlement contract (from the withdrawal contract) and to do the final payout to the order owners. This would allow us to save some gas in the order processing by not having to allocate large `Orders[] memory` array (since an order is right now ~3x larger as memory structs aren't tightly packed AFAIK).

Reply:
What we could do is use the decoding functions for all checks (nonce, limit price, ...?), and only output the transfer operations that are expected for this order (in and out), only storing that in memory. As far as I can think right now, this is the only part that can't be used on the spot while decoding. (Maybe you can also execute the withdrawals when decoding, depending on if the price of the extra call for each order outweighs the price of storing all transfers simultaneously in memory.)

Reply:
ubernit: instead of returning the digest, can we add it to the order struct (so we don't mix side-effects and return types)?

Reply:
> so we don't mix side-effects and return types

I'm not sure I understand what you mean?

Reply:
So the main reason I split it this way was that the `Order memory` gets filled with order details, and we return the intermediate values (such as `digest` and `*TokenIndex`) that aren't actually part of the order but may be useful to the caller to:
- compute the nonce storage slot
- verify limit prices *without* searching the token list.

Reply:
> I'm not sure I understand what you mean?

By side effect I mean populating the order struct. In the current version we are returning some values as part of the return type and some values by populating a reference. I was suggesting to put all values on the Order type as I thought you are not planning to pass this one around anyways (but instead only keep `(token: address, recipient: address, executedSellAmount: uint256, executedBuyAmount: uint256)`). But I might not understand what the different types are used for (it just seems odd to me to have both ways of returning data mixed).

Reply:
> I was suggesting to put all values on the Order type as I thought you are not planning to pass this one around anyways

That's a very good point. I can change this.